sqld/src/replication/replica/injector/mod.rs

```diff
@@ -0,0 +1,220 @@
+            capacity: buffer_capacity,
+            connection: Arc::new(Mutex::new(connection)),
+            _hook_ctx: ctx,
+        })
+    }
+
+    /// Inject on frame into the log. If this was a commit frame, returns Ok(Some(FrameNo)).
+    pub(crate) fn inject_frame(&mut self, frame: Frame) -> crate::Result<Option<FrameNo>> {
+        let frame_close_txn = frame.header().size_after != 0;
+        self.buffer.lock().push_back(frame);
+        if frame_close_txn || self.buffer.lock().len() >= self.capacity {
+            if !self.is_txn {
+                self.begin_txn()?;
+            }
+            return self.flush();
+        }
+
+        Ok(None)
+    }
+
+    /// Flush the buffer to libsql WAL.
+    /// Trigger a dummy write, and flush the cache to trigger a call to xFrame. The buffer's frame
+    /// are then injected into the wal.
+    fn flush(&mut self) -> crate::Result<Option<FrameNo>> {
+        let lock = self.buffer.lock();
+        // the frames in the buffer are either monotonically increasing (log) or decreasing
+        // (snapshot). Either way, we want to find the biggest frameno we're about to commit, and
+        // that is either the front or the back of the buffer
+        let last_frame_no = match lock.back().zip(lock.front()) {
+            Some((b, f)) => f.header().frame_no.max(b.header().frame_no),
+            None => {
+                tracing::trace!("nothing to inject");
+                return Ok(None);
+            }
+        };
+
+        drop(lock);
+
+        let connection = self.connection.lock();
+        connection.execute("INSERT INTO __DUMMY__ VALUES (42)", ())?;
+        // force call to xframe
+        match connection.cache_flush() {
+            Ok(_) => panic!("replication hook was not called"),
+            Err(e) => {
+                if let Some(e) = e.sqlite_error() {
+                    if e.extended_code == LIBSQL_INJECT_OK {
+                        // refresh schema
+                        connection.pragma_update(None, "writable_schema", "reset")?;
+                        let _ = connection.execute("COMMIT", ()); // TODO: Handle error?
+                        self.is_txn = false;
+                        assert!(self.buffer.lock().is_empty());
+                        return Ok(Some(last_frame_no));
+                    } else if e.extended_code == LIBSQL_INJECT_OK_TXN {
+                        self.is_txn = true;
+                        assert!(self.buffer.lock().is_empty());
+                        return Ok(None);
+                    } else if e.extended_code == LIBSQL_INJECT_FATAL {
+                        todo!("handle fatal error");
+                    }
+                }
+
+                todo!("handle fatal error");
+            }
+        }
+    }
+
+    fn begin_txn(&mut self) -> crate::Result<()> {
+        let conn = self.connection.lock();
+        conn.execute("BEGIN IMMEDIATE", ())?;
+        conn.execute("CREATE TABLE __DUMMY__ (__dummy__)", ())?;
```

Comment:
Do I understand correctly that this table will show up in user's schemas now? If so, we need it prefixed with `libsql_`, and documented. Also, the previous code just called `CREATE TABLE __DUMMY__(__dummy__)` in order to trigger xFrames -- why is the new impl different?

Reply:
The injection mechanism is quite different now. Before, we had a call to `CREATE TABLE __DUMMY__ ` drop us in the `x_frame` context. From there, we had a channel pass the frame from the replication fiber to the connection fiber. The call was aborted, rolling back the txn by returning an error.
With the new method, we want to re-enter the x-frame call multiple times from within the same transaction context. We therefore first create the dummy table, and then _pretend_-add multiple values in it. Note that this does not necessarily trigger a call to `x_frame`. This is what `cache_flush` does, and forces a call to x_frame. when we have written all of our frames for this transaction, we call `writable_schema`, to restore the in-memory schema from disk (removing the dummy table from the in memory schema), and then commit.

Reply:
Ok, I'd need to go over this new algorithm more closely. I think we're going to either:
1. Call this temporary table `libsql__dummy__`, to make it as unlikely to clash as possible - SQLite uses `sqlite_` prefix for internal tables, and we already use `libsql_` prefix for the Wasm functions table
2. Just create a `libsql_frame_injection_dummy` table in sqld and assume it exists. Then we don't need to keep modifying the schema twice on each injection, making it more robust.

I'm strongly in favor of (2.), unless we have a good reason not to do it. What do you think